### PR TITLE
PERF: BaseMaskedArray.__iter__

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Faster iteration through EA's backed by BaseMaskedArray.

ASV added for iterating through a `Series` backed by various dtypes:

```
       before           after         ratio
     [b6736b44]       [b9193932]
                      <masked-array-iter>
-      16.8±0.9ms      5.14±0.06ms     0.31  series_methods.Iter.time_iter('Float64')
-        17.2±1ms       5.15±0.1ms     0.30  series_methods.Iter.time_iter('Int64')
-        16.8±1ms      3.50±0.02ms     0.21  series_methods.Iter.time_iter('boolean')
```

### CODE REVIEW

File: pandas/core/arrays/masked.py [hunk 1]
```diff
@@ -247,11 +247,16 @@ def __setitem__(self, key, value) -> None:
 
     def __iter__(self) -> Iterator:
         if self.ndim == 1:
```
Comment (jbrockmendel) [inline pandas/core/arrays/masked.py:249]: What if I set an NAvalue during iteration?

Comment (jbrockmendel) [reply pandas/core/arrays/masked.py:249]: I guess a lot of iter methods wouldn’t be robust to that. Not worth worrrying about

Comment (lukemanley) [reply pandas/core/arrays/masked.py:249]: If you think its worth handling that case, I can remove the `if` block that iterates without the mask. The `else` block is robust to your example and we still get a perf improvement if we were to hit `'else` block for everything:

```
import pandas as pd
import numpy as np

arr_without_na = pd.array(np.arange(10**6), dtype="Int64")

arr_with_na = arr_without_na.copy()
arr_with_na[::2] = pd.NA

def time_iter(arr):
    for v in arr:
        pass

%timeit time_iter(arr_without_na)
%timeit time_iter(arr_with_na)

# without NA's
214 ms ± 13.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     <- main
56.3 ms ± 1.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  <- PR

# with NA's
273 ms ± 27.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     <- main
85.6 ms ± 902 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)   <- PR
```

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -601,7 +601,7 @@ Performance improvements
 - Performance improvement in :meth:`DataFrame.join` when joining on a subset of a :class:`MultiIndex` (:issue:`48611`)
 - Performance improvement for :meth:`MultiIndex.intersection` (:issue:`48604`)
 - Performance improvement in ``var`` for nullable dtypes (:issue:`48379`).
-- Performance improvement when iterating over a :class:`~arrays.ArrowExtensionArray` (:issue:`49825`).
+- Performance improvement when iterating over various extension array types (:issue:`49825`, :issue:`49851`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:604]:
```suggestion
- Performance improvement when iterating over pyarrow and nullable dtype (:issue:`49825`, :issue:`49851`)
```

File: pandas/core/arrays/masked.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/masked.py:249]: no not worth worrying about

File: doc/source/whatsnew/v2.0.0.rst (hunk 2 above)
Comment (lukemanley) [reply doc/source/whatsnew/v2.0.0.rst]: updated, thanks

**Review by mroeschke — APPROVED** [review]
LGTM cc @jbrockmendel merge when ready

**Pull request merged**